File: leetcode/95.不同的二叉搜索树 II/95.不同的二叉搜索树 II.py
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def generateTrees(self, n: int) -> List[TreeNode]:
        if n == 0:
            return []
        return self.generate(1, n)

    def generate(self, low, high) -> List[TreeNode]:
        res = []
        if low > high:
            res.append(None)
            return res
        for i in range(low, high + 1):
            # 递归构造左右子树
            left = self.generate(low, i - 1)
            right = self.generate(i + 1, high)
            # 给root节点穷举所有左右子树
            for l in left:
                for r in right:
                    root = TreeNode(i)
                    root.left = l
                    root.right = r
                    res.append(root)
        return res

File: leetcode/95.不同的二叉搜索树 II/95.不同的二叉搜索树 II.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func generateTrees(n int) []*TreeNode {
	if n == 0 {
		return []*TreeNode{}
	}
	return generate(1, n)
}
func generate(low int, high int) []*TreeNode {
	res := []*TreeNode{}
	// base case 无区间时，返回nil节点
	if low > high {
		res = append(res, nil)
		return res
	}
	for i := low; i <= high; i++ {
		// 递归构建左右子树
		left := generate(low, i-1)
		right := generate(i+1, high)
		// 给root节点穷举所有左右子树
		for _, l := range left {
			for _, r := range right {
				root := &TreeNode{Val: i}
				root.Left = l
				root.Right = r
				res = append(res, root)
			}
		}
	}
	return res
}
